### ISSUE TITLE
reveal-ck senses the update of index.html and rebuilds indefinitely

### ISSUE THREAD

**Issue by agate-pris** [jedcn/reveal-ck#98]
# Overview

Perhaps specifying a deep hierarchical file for the `--file` option causes an infinite loop on `rebuild` at `serve`. I think the same thing happens when the `--dir` option is specified.

# Proposal

Instead of monitoring files other than `slides/index.html` in the work directory, monitor the directory where the input source file exists and the output destination directory except `index.html`.

The reason for monitoring the output destination directory is because it is expected that image files etc will be placed in the output destination directory.

The reason for monitoring the directory where the input source file placed is because it is unnatural to use the file above the input source file.

The reason why should not monitor the work directory is that when working on a shallow layer of the `C` drive etc, it cause performance issue on `guard` and `guard-livereload`.

# Detailed logs

```
PS D:\test> reveal-ck --version
reveal-ck version 3.9.0
PS D:\test> reveal-ck generate -f ./foo/bar.md
Generating slides for './foo/bar.md'..
PS D:\test> reveal-ck serve -f ./foo/bar.md
------------------------------------------------------------
[ reveal-ck ] Serving up slide content in './foo/bar/'.
[ reveal-ck ] Open your browser to 'http://localhost:10000'.
[ reveal-ck ] Press CTRL-C to stop.
------------------------------------------------------------
[ reveal-ck ] Getting Ready to Reload Browsers.
[ reveal-ck ] Getting Ready to Rebuild Slides.
[ reveal-ck ] Starting Webserver.
[   reload  ] Using Guardfile at C:/tools/ruby25/lib/ruby/gems/2.5.0/gems/reveal-ck-3.9.0/files/reveal-ck/Guardfile.
[   reload  ] Run 'gem install win32console' to use color on Windows
[   reload  ] LiveReload is waiting for a browser to connect.
[   reload  ] Guard is now watching at 'D:/test'
[   reload  ] Browser connected.
[  rebuild  ] Modified: D:/test/foo/bar.md
[  rebuild  ] Generating slides for './foo/bar.md'..
[  rebuild  ] Modified: D:/test/foo/bar.md, D:/test/foo/bar/slides.html, D:/test/foo/bar/index.html
[  rebuild  ] Generating slides for './foo/bar.md'..
[  rebuild  ] Modified: D:/test/foo/bar/slides.html, D:/test/foo/bar/index.html
[  rebuild  ] Generating slides for './foo/bar.md'..
[  rebuild  ] Modified: D:/test/foo/bar/slides.html, D:/test/foo/bar/index.html
[  rebuild  ] Generating slides for './foo/bar.md'..
```

The same line continued infinitely from the above.

It was same on Git Bash for Windows as well as Windows PowerShell.

I opened `C:\tools\ruby25\lib\ruby\gems\2.5.0\gems\reveal-ck-3.9.0\files\reveal-ck\Guardfile`.
The contents are as follows.

```
notification :off

guard 'livereload' do
  watch(%r{^slides/index.html$})
end

```

**Comment by agate-pris** [comment]
j

**Comment by agate-pris** [comment]
I'm sorry, closed by misoperation.

**Comment by jedcn** [comment]
Hey there, thanks for the well written issue!

I'll take a look shortly.

**Comment by jedcn** [comment]
There is a problem here but I'm not seeing an infinite rebuild.

Instead I get a situation where no changes are detected and so the browser is not reloaded.

The issue appears to be that [this watch line of the Guardfile][guardfile-listen] is hardcoded to `slides/` when it should instead be the dynamic value of `output_dir` from `serve.rb` [over here][serve-rb-output-dir].

[guardfile-listen]: https://github.com/jedcn/reveal-ck/blob/master/files/reveal-ck/Guardfile#L4
[serve-rb-output-dir]: https://github.com/jedcn/reveal-ck/blob/master/lib/reveal-ck/commands/serve.rb#L18

Here are my repro steps (mac, not windows-- could that be an issue?)

```
cd /tmp
mkdir foo
echo "First Slide" >> foo/bar.md
reveal-ck generate -f foo/bar.md
reveal-ck serve -f foo/bar.md
```

Then open a browser to http://localhost:10000, you'll see "First Slide"

Then modify `/tmp/foo/bar.md` so that it has "First Slide!" (a new exclamation point), I'd expect it'd reload but instead it does not reload. The terminal output actually looks like:

```
[  rebuild  ] Modified: /private/tmp/foo/bar.md
[  rebuild  ] Generating slides for 'foo/bar.md'..
[  rebuild  ] Modified: /private/tmp/foo/bar/index.html, /private/tmp/foo/bar/slides.html
```

And for each change I make and save I see this-- the slides are generated once for each save but the browser is not reloaded.

It looks like your output:

```
[  rebuild  ] Modified: /private/tmp/foo/bar.md
[  rebuild  ] Generating slides for 'foo/bar.md'..
[  rebuild  ] Modified: /private/tmp/foo/bar/index.html, /private/tmp/foo/bar/slides.html
[  rebuild  ] Generating slides for 'foo/bar.md'..
[  rebuild  ] Modified: /private/tmp/foo/bar/index.html, /private/tmp/foo/bar/slides.html
[  rebuild  ] Generating slides for 'foo/bar.md'..
```

However there is no infinite loop. Each of those three lines appears whenever I make  change and save. But if I stop saving it the lines stop appearing.

When I'd expect to see would include a reload portion. When I run from development instead of the 3.9.0 gem and manually change the `Guardfile` to be correct each save looks like:

```
[  rebuild  ] Generating slides for 'foo/bar.md'..
[  rebuild  ] Modified: /projects/reveal-ck/foo/bar/index.html, /projects/reveal-ck/foo/bar/slides.html
[  rebuild  ] Generating slides for 'foo/bar.md'..
[   reload  ] Reloading browser: foo/bar/index.html foo/bar/index.html
```

Where the key line is `[   reload  ] Reloading browser: foo/bar/index.html foo/bar/index.html`

---

If you're seeing an infinite loop then perhaps something is different from Windows to Mac?

I'll do some digging around.

In the mean time-- you should be able to use reveal-ck with a flat (non-nested) directory structure like so:

Don't setup:

* `foo/bar.md`, `foo/baz.md`, `foo/boop.md`

Do setup:

* `foo/bar/slides.md`, `foo/baz/slides.md`, `foo/boop/slides.md`

And cd into each of `foo/bar`, `foo/baz`, and `foo/boop` when running `generate` or `serve` commands.

Does this suggestion work for you?

**Comment by agate-pris** [comment]
I changed `Guardfile` to some patterns.

First,

```rb
notification :off

guard 'livereload' do
  watch(%r{^foo/bar/index.html$})
end
```

second,

```rb
notification :off

guard 'livereload' do
  watch(%r{foo/bar.md$})
end
```

and thrid,

```rb
notification :off

guard 'livereload' do
end
```

then infinite rebuilds occurred in all patterns.

I placed `slides.md` at current directory and exec `reveal-ck generate` and `reveal-ck serve` with no option with third `Guardfile` pattern.

I modified `slides.md` then `rebuild` has run and noticed that `reload` has **not** run.

Is it different from rebuild file change detection and reload file change detection? It seemed to me that the `rebuild` process did not use `Guardfile`.

I might want to write "Do not use `dir` and `file` option" to `README.md` for a while because it seems to behave counterintuitive.

**Comment by jedcn** [comment]
Thank you for all of your hard work!

There are two actions that reveal-ck takes when it learns of a change--

1. rebuilding slides
2. reloading the browser

Both of them are hard coded to use `slides` and must be changed.

1. [Here is the part that listens and rebuilds slides][rebuild-slides].
2. [Here is the part that listens and reloads the browser][reload-browser].

[rebuild-slides]: https://github.com/jedcn/reveal-ck/blob/master/lib/reveal-ck/commands/listen_to_rebuild_slides.rb#L30
[reload-browser]: https://github.com/jedcn/reveal-ck/blob/master/lib/reveal-ck/commands/listen_to_reload_browser.rb#L23

I will have some time to look into this today and tomorrow. I hope to publish a branch or a new version with changes that'll help.

**Comment by agate-pris** [comment]
Thank you, jedcn.

I could find the code aboud `reload`, but I could not find the code about `rebuild` (Since I looked at the log on the terminal and searched for the file name by looking up it).

Looking at the behavior, I thought that it was probably detecting in a different way, but I did not think that the content was hard-coded on the ruby code.

> I will have some time to look into this today and tomorrow. I hope to publish a branch or a new version with changes that'll help.

Not only quick reply, it's very cool to fix it too quickly!
